extern/boostd-data/bench/cassandra.go

```diff
@@ -0,0 +1,230 @@
+		if err != nil {
+			return fmt.Errorf("creating tables: executing\n%s\n%w", line, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CassandraDB) Cleanup(ctx context.Context) error {
+	_ = c.session.Query(`drop table if exists bench.PayloadToPieces`).WithContext(ctx).Exec()
+	_ = c.session.Query(`drop table if exists bench.PieceBlockOffsetSize`).WithContext(ctx).Exec()
+	_ = c.session.Query(`drop keyspace bench`).WithContext(ctx).Exec()
+	c.session.Close()
+	return nil
+}
+
+func (c *CassandraDB) GetBlockSample(ctx context.Context, count int) ([]pieceBlock, error) {
+	// TODO: randomize order
+	qry := `SELECT PieceCid, PayloadMultihash FROM bench.PieceBlockOffsetSize LIMIT ?`
+	iter := c.session.Query(qry, count).WithContext(ctx).Iter()
+
+	var pieceCidBz, payloadMHBz []byte
+	pbs := make([]pieceBlock, 0, count)
+	for iter.Scan(&pieceCidBz, &payloadMHBz) {
+		_, pcid, err := cid.CidFromBytes(pieceCidBz)
+		if err != nil {
+			return nil, fmt.Errorf("scanning piece cid: %w", err)
+		}
+		_, pmh, err := multihash.MHFromBytes(payloadMHBz)
+		if err != nil {
+			return nil, fmt.Errorf("scanning mulithash: %w", err)
+		}
+
+		pbs = append(pbs, pieceBlock{
+			PieceCid:         pcid,
+			PayloadMultihash: pmh,
+		})
+	}
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+
+	//log.Debug("got pbs:")
+	//for _, pb := range pbs {
+	//	log.Debugf("  %s %s", pb.PieceCid, pb.PayloadMultihash)
+	//}
+
+	return pbs, nil
+}
+
+func (c *CassandraDB) AddIndexRecords(ctx context.Context, pieceCid cid.Cid, recs []model.Record) error {
+	if len(recs) == 0 {
+		return nil
+	}
+
+	batchEntries := make([]gocql.BatchEntry, 0, 2*len(recs))
+
+	// Add payload to pieces index
+	for _, rec := range recs {
+		batchEntries = append(batchEntries, gocql.BatchEntry{
+			Stmt:       `INSERT INTO bench.PayloadToPieces (PayloadMultihash, PieceCids) VALUES (?, ?)`,
+			Args:       []interface{}{rec.Cid.Hash(), pieceCid.Bytes()},
+			Idempotent: true,
+		})
+	}
+
+	// Add piece to block info index
+	for _, rec := range recs {
+		batchEntries = append(batchEntries, gocql.BatchEntry{
+			Stmt:       `INSERT INTO bench.PieceBlockOffsetSize (PieceCid, PayloadMultihash, BlockOffset, BlockSize) VALUES (?, ?, ?, ?)`,
+			Args:       []interface{}{pieceCid.Bytes(), rec.Cid.Hash(), rec.Offset, rec.Size},
+			Idempotent: true,
+		})
+	}
+
+	// Cassandra has a 50k limit on batch statements. Keeping batch size small
+	// makes sure we're under the limit.
+	const batchSize = 128
+	var batchIdx int
+	var batch *gocql.Batch
+	for allIdx, entry := range batchEntries {
+		if batchIdx == 0 {
+			batch = c.session.NewBatch(gocql.UnloggedBatch).WithContext(ctx)
+		}
+		if allIdx == len(batchEntries)-1 || batchIdx == batchSize {
+			err := c.session.ExecuteBatch(batch)
+			if err != nil {
+				return fmt.Errorf("adding index records for piece %s: %w", pieceCid, err)
+			}
+			batchIdx = 0
+			continue
+		}
+
+		batch.Entries = append(batch.Entries, entry)
+		batchIdx++
```

Comment:
This code is moderately confusing, to me this reads like the last batch entry will never be it a batch that is executed.

The append should probably be before this if.